src/main/java/org/folio/rest/impl/LoginAPI.java

```diff
@@ -199,31 +232,225 @@ private Future<String> fetchToken(JsonObject payload, String tenant,
+        .build();
+  }
+
+  private Response tokenResponseLegacy(JsonObject tokenJson) {
+    String authToken = tokenJson.getString("token");
+    var response = new LoginResponse().withOkapiToken(authToken);
+    return PostAuthnLoginResponse.respond201WithApplicationJson(response,
+        PostAuthnLoginResponse.headersFor201().withXOkapiToken(authToken));
+  }
+
+  private void handleLogout(String cookieHeader, Map<String, String> okapiHeaders,
+  Handler<AsyncResult<Response>> asyncResultHandler) {
+    try {
+      logger.debug("Cookie in handleLogout is: {}", cookieHeader);
+
+      String tenantId = getTenant(okapiHeaders);
+      String okapiURL = okapiHeaders.get(XOkapiHeaders.URL);
+      String okapiToken = okapiHeaders.get(XOkapiHeaders.TOKEN);
+
+      Future<HttpResponse<Buffer>> logoutFuture;
+      if (cookieHeader == null) {
+        logoutFuture = logoutAll(tenantId, okapiURL, okapiToken);
+      } else {
+        var p = new TokenCookieParser(cookieHeader);
+        logoutFuture = logout(tenantId, okapiURL, okapiToken, p.getRefreshToken());
+      }
+
+      logoutFuture.onSuccess(r -> {
+      if (r.statusCode() >= 400) {
+          logger.error("{} (status code: {})", TOKEN_LOGOUT_UNPROCESSABLE, r.statusCode());
+          asyncResultHandler.handle(Future.succeededFuture(
+            PostAuthnLogoutResponse.respond422WithApplicationJson(getErrors(
+              TOKEN_LOGOUT_UNPROCESSABLE, TOKEN_LOGOUT_UNPROCESSABLE_CODE))));
+          return;
+        }
+
+        asyncResultHandler.handle(Future.succeededFuture(logoutResponse()));
+      });
+
+      logoutFuture.onFailure(e -> {
+        logger.error("{}: {}", INTERNAL_ERROR, e.getMessage(), e);
+        asyncResultHandler.handle(Future.succeededFuture(PostAuthnLogoutResponse.respond500WithTextPlain(
+          getErrors(INTERNAL_ERROR, TOKEN_LOGOUT_FAIL_CODE))));
+      });
+    } catch (Exception e) {
+      String msg = "Unexpected exception when handling logout";
+      logger.error("{}: {}", msg, e.getMessage(), e);
+      asyncResultHandler.handle(Future.succeededFuture(PostAuthnLogoutResponse.respond500WithTextPlain(INTERNAL_ERROR)));
+    }
+  }
+
+  @Override
+  public void postAuthnLogout(String cookieHeader, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context ctx) {
+    handleLogout(cookieHeader, okapiHeaders, asyncResultHandler);
+  }
+
+  @Override
+  public void postAuthnLogoutAll(Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context ctx) {
+    handleLogout(null, okapiHeaders, asyncResultHandler);
+  }
+
+  @Override
+  public void postAuthnRefresh(String cookieHeader, Map<String, String> otherHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context ctx) {
+    logger.debug("Cookie is {}", cookieHeader);
+
+    String refreshToken = "";
+    try {
+      var p = new TokenCookieParser(cookieHeader);
```

Comment:
This parser fails if access token is not part of Cookie. Is it the intention that refreshing requires an access token?

Reply:
I think you are right. I'll fix this.